### PR TITLE
Run a simulation against a hub provisioned by the user

### PR DESCRIPTION
**Pull request by jillcary**
# Type of change? 
- [ ] Bug fix
- [x] New feature
- [ ] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 

- Add new field for IotHub ConnectionString to SimulationApiModel
- When user provides a connection string:
  - the value is stored to a file on the vm
  - the key is removed from the connection string and the string is returned to the API model
  - checks if the connection string is a valid iot hub
  - checks that the cconnection string is a valid format
- the caller can specify "pre-provisioned" to switch back to the value stored in PCS_IOTHUB_CONNSTRING

**Checklist:**

- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [x] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: device-simulation.sln [hunk 1]
```diff
@@ -1,17 +1,19 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
```
Comment (dluc) [inline device-simulation.sln]: unless we need these changes, could you revert them?

File: WebService/v1/Models/IotHubApiModel.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        private const string USE_LOCAL_IOTHUB = "pre-provisioned";
+        private string iotHubConnectionString;
+
+        [JsonProperty(PropertyName = "ConnectionString")]
+        public string ConnectionString
+        {
+            get => this.iotHubConnectionString;
+            set =>
+                // remove and securely store senstive key information
```
Comment (dluc) [inline WebService/v1/Models/IotHubApiModel.cs]: typo

File: WebService/v1/Models/Helpers/IotHubConnectionStringManager.cs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+using System.IO;
+using System.Text.RegularExpressions;
+using Microsoft.Azure.Devices;
+using Microsoft.Azure.Devices.Common;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.WebService.v1.Exceptions;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.WebService.v1.Models.Helpers
+{
+    public class IotHubConnectionStringManager
```
Comment (dluc) [inline WebService/v1/Models/Helpers/IotHubConnectionStringManager.cs]: I would move this class in the service layer

File: WebService/v1/Models/Helpers/IotHubConnectionStringManager.cs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    {
+        private const string USE_LOCAL_IOTHUB = "pre-provisioned";
+        private const string CONNSTRING_FILE_PATH = @"keys\user_iothub_key.txt";
```
Comment (dluc) [inline WebService/v1/Models/Helpers/IotHubConnectionStringManager.cs]: file paths should be cross platform, there is constant you can use for folder separator

Comment (dluc) [reply WebService/v1/Models/Helpers/IotHubConnectionStringManager.cs]: the path might not exist, perhaps write in the "current" folder instead?

File: WebService/v1/Exceptions/InvalidIotHubConnectionStringFormatException.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.WebService.v1.Exceptions
+{
+    public class InvalidIotHubConnectionStringFormatException : Exception
+    {
+        /// <summary>
+        /// This exception is thrown when a the IoTHub connection string provided
```
Comment (dluc) [inline WebService/v1/Exceptions/InvalidIotHubConnectionStringFormatException.cs]: typo

File: Services/Models/Simulation.cs [hunk 6]
```diff
@@ -37,6 +40,8 @@ public Simulation()
 
             // When unspecified, a simulation is enabled
             this.Enabled = true;
+            // by default, use PCS_IOTHUB_CONNSTRING
```
Comment (dluc) [inline Services/Models/Simulation.cs]: "...env var"

File: Services/IotHub/IotHubConnectionStringManager.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+using Microsoft.Azure.Devices.Common;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Runtime;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.IotHub
+{
+    // retrieves Iot Hub connection secret from storage
+    public class IotHubConnectionStringManager
+    {
+        private const string CONNECTION_STRING_FILE_PATH = @"keys\user_iothub_key.txt";
```
Comment (dluc) [inline Services/IotHub/IotHubConnectionStringManager.cs]: the value is duplicated in IotHubConnectionStringManager, we should store similar constants in one place

File: Services/IotHub/IotHubConnectionStringManager.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        /// </summary>
+        /// <param name="path"></param>
+        /// <returns></returns>
+        private static string ReadFromFile(string path)
+        {
+            if (File.Exists(path))
+            {
+                // remove special characters and return string
+                return Regex.Replace(File.ReadAllText(path), @"[\r\n\t ]+", "");
+            }
```
Comment (dluc) [inline Services/IotHub/IotHubConnectionStringManager.cs]: else... log info "file not present"

Comment (jillcary) [reply Services/IotHub/IotHubConnectionStringManager.cs]: Done.

File: docs/API_SPECS_SIMULATIONS.md [hunk 9]
```diff
@@ -123,6 +123,40 @@ Content-Type: application/json; charset=utf-8
...
+
+To specify a custom Azure IoT Hub, specify the connection string for the IoT Hub in the PUT request.
+Ex: `"ConnectionString": "HostName=iothub-abcde.azure-devices.net;SharedAccessKeyName=iothubowner;SharedAccessKey=<valid iothub key>"`
+
+The response will contain the connection string with the sensitive key information removed.
+
+If a valid connection string has already been provided, a partial connection string with a null key may be provided.
+Ex: `"ConnectionString": "HostName=iothub-abcde.azure-devices.net;SharedAccessKeyName=iothubowner;SharedAccessKey="`
+
+To switch back to the default Azure IoT Hub stored in the config file, specficy "default" as the connection string.
```
Comment (dluc) [inline docs/API_SPECS_SIMULATIONS.md]: typo

**Pull request merged**